### PR TITLE
Fix: PM-blocking a player does not stop them inviting you to a party or guild, fixes #8190

### LINKED ISSUE #8190: PM-blocking a player does not stop them inviting you to a party or guild
When a player uses the "block" feature on another user's profile to prevent that user from sending a private message to the player, the user is still able to invite the player to a party or guild. This is undesirable because blocking a user from sending you PMs should prevent all contact from that user.

This is the PM block feature that I'm referring to (see the icon at the bottom with the "Block" on-hover text):

image

### PR DESCRIPTION
**Pull request by rrgayhart**
Fixes #8190 

### Changes

- Fixes issue where blocking a player does not stop them from inviting you to a party of guild
- Moves shared logic of detecting a blockedUser to it's own function
- Adds unit and integration testing

----
UUID: 4f82317e-6f0c-4855-9512-3b9b8ca0017d

### CODE REVIEW

File: website/common/script/libs/userBlockPresent.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module.exports = function userBlockPresent (sender, receiver) {
+  let userBlockedSender = receiver.inbox.blocks.indexOf(sender._id) !== -1;
```
Comment (crookedneighbor) [inline website/common/script/libs/userBlockPresent.js]: Let's make this a bit more DRY

``` js
function isBlocked(user1, user2) {
  return user1.inbox.blocks.indexOf(user2._id) !== -1
}

module.exports = function userBlockPresent (sender, receiver) {
  let userBlockedSender = isBlocked(receiver, sender);
  let senderBlockedReceiver = isBlocked(sender, receiver);
  return userBlockedSender || senderBlockedReceiver;
};
```

**Review by crookedneighbor — CHANGES_REQUESTED** [review]

**Comment by crookedneighbor** [comment]
Overall, seems reasonable.

File: website/common/script/libs/userBlockPresent.js (hunk 1 above)
Comment (khipkin) [reply website/common/script/libs/userBlockPresent.js]: I agree! But to be fair, the code was like this originally. 😛

Comment (crookedneighbor) [reply website/common/script/libs/userBlockPresent.js]: Picard Nodding

**Comment by crookedneighbor** [comment]
Looks to me like this isn't used anywhere on the client. We may want to implement this logic later on the client, but for now it should be moved to a server lib instead.

File: website/common/script/libs/userBlockPresent.js (hunk 1 above)
Comment (rrgayhart) [reply website/common/script/libs/userBlockPresent.js]: 😭 ❤️

**Comment by rrgayhart** [comment]
@crookedneighbor Changes implemented! LMK if you see additional adjustments!

File: test/api/v3/integration/groups/POST-groups_invite.test.js [hunk 2]
```diff
@@ -97,6 +97,19 @@ describe('Post /groups/:groupId/invite', () => {
       });
     });
 
+    it('returns error when to receiver has blocked the inviter', async () => {
```
Comment (khipkin) [inline test/api/v3/integration/groups/POST-groups_invite.test.js]: "the" receiver 😈

Comment (rrgayhart) [reply test/api/v3/integration/groups/POST-groups_invite.test.js]: 😱 Haha - will fix right now!

**Comment by rrgayhart** [comment]
@crookedneighbor - Quick ping reminder that this PR is ready to be re-looked at when you're able!

File: website/server/controllers/api-v3/members.js [hunk 3]
```diff
@@ -306,11 +307,11 @@ api.sendPrivateMessage = {
...
     if (!receiver) throw new NotFound(res.t('userNotFound'));
 
-    let userBlockedSender = receiver.inbox.blocks.indexOf(sender._id) !== -1;
-    let userIsBlockBySender = sender.inbox.blocks.indexOf(receiver._id) !== -1;
     let userOptedOutOfMessaging = receiver.inbox.optOut;
 
-    if (userBlockedSender || userIsBlockBySender || userOptedOutOfMessaging) {
+    let userBlock = userBlockPresent(sender, receiver);
+
+    if (userBlock || userOptedOutOfMessaging) {
```
Comment (TheHollidayInn) [inline website/server/controllers/api-v3/members.js:314]: You should add a new test for this route.

**Comment by TheHollidayInn** [comment]
I'm a little late to the review, but I added one comment. Overall, this looks really good. Your coding style is nice.

**Comment by SabreCat** [comment]
This covers a lot of the same ground as https://github.com/HabitRPG/habitica/commit/018976a7231683420acd982bf9705b0046846840 ; I think for this we should expand that functionality to cover these scenarios rather than implement a parallel check. Closing this for now.

**Pull request closed without merging**